src/st_stuff.c

```diff
@@ -86,1400 +84,1712 @@
+    }
+    else
+    {
+        number->xoffset += width;
+    }
+}
+
+static void DrawGlyphLine(int x, int y, sbarelem_t *elem, widgetline_t *line,
+                          patch_t *glyph)
+{
+    sbe_widget_t *widget = elem->pointer.widget;
+    hudfont_t *font = widget->font;
+
+    int width, widthdiff;
+
+    if (font->type == sbf_proportional)
+    {
+        width = glyph ? SHORT(glyph->width) : SPACEWIDTH;
+        widthdiff = 0;
+    }
     else
-      break;
-  }
-  have_xdthfaces = i;
+    {
+        width = font->monowidth;
+        widthdiff = glyph ? SHORT(glyph->width) - width : 0;
+    }
+
+    if (elem->alignment & sbe_h_middle)
+    {
+        line->xoffset += ((width + widthdiff) >> 1);
+    }
+    else if (elem->alignment & sbe_h_right)
+    {
+        line->xoffset += (width + widthdiff);
+    }
+
+    if (glyph)
+    {
+        DrawPatch(x + line->xoffset, y, elem->alignment, glyph, elem->cr,
+                  elem->tranmap);
+    }
+
+    if (elem->alignment & sbe_h_middle)
+    {
+        line->xoffset += (width - ((width - widthdiff) >> 1));
+    }
+    else if (elem->alignment & sbe_h_right)
+    {
+        line->xoffset += -widthdiff;
+    }
+    else
+    {
+        line->xoffset += width;
+    }
 }
 
-void ST_loadData(void)
+static void DrawNumber(int x, int y, sbarelem_t *elem)
 {
-  lu_palette = W_GetNumForName ("PLAYPAL");
-  ST_loadGraphics();
+    sbe_number_t *number = elem->pointer.number;
+
+    int value = number->value;
+    int base_xoffset = number->xoffset;
+    numberfont_t *font = number->font;
+
+    if (value < 0 && font->minus != NULL)
+    {
+        DrawGlyphNumber(x, y, elem, font->minus);
+        value = -value;
+    }
+
+    int glyphindex = number->numvalues;
+    while (glyphindex > 0)
+    {
+        int glyphbase = (int)pow(10.0, --glyphindex);
+        int workingnum = value / glyphbase;
+        DrawGlyphNumber(x, y, elem, font->numbers[workingnum]);
+        value -= (workingnum * glyphbase);
+    }
+
+    if (elem->type == sbe_percent && font->percent != NULL)
+    {
+        crange_idx_e oldcr = elem->crboom;
+        if (sts_pct_always_gray)
+        {
+            elem->crboom = CR_GRAY;
+        }
+        DrawGlyphNumber(x, y, elem, font->percent);
+        elem->crboom = oldcr;
+    }
+
+    number->xoffset = base_xoffset;
 }
 
-void ST_initData(void)
+static void DrawLines(int x, int y, sbarelem_t *elem)
 {
-  int i;
+    sbe_widget_t *widget = elem->pointer.widget;
 
-  st_firsttime = true;
-  plyr = &players[displayplayer];            // killough 3/7/98
+    int cr = elem->cr;
 
-  st_statusbaron = true;
+    widgetline_t *line;
+    array_foreach(line, widget->lines)
+    {
+        int base_xoffset = line->xoffset;
+        hudfont_t *font = widget->font;
 
-  st_faceindex = 0;
-  st_palette = -1;
+        const char *str = line->string;
+        while (*str)
+        {
+            int ch = *str++;
 
-  st_oldhealth = -1;
+            if (ch == '\x1b' && str)
```

Comment:
`&& *str` ?

Reply:
Fixed in https://github.com/fabiangreffrath/woof/pull/1916/commits/deb884d89cfc874bfb4bd09bdad259575cf06a14, thanks! Perhaps one day we should consolidate all the string drawing functions.

Reply:
There is another occurrence in line 758.

Reply:
Fixed here: https://github.com/fabiangreffrath/woof/pull/1916/commits/78488d25687334499b960158b92d416e495cb204